src/Modules/UngradedModules/ModuleGens.jl

```diff
@@ -263,47 +263,44 @@ Convert a Singular vector to a free module element.
 """
 function (F::FreeMod{<:MPolyRingElem})(s::Singular.svector)
   Rx = base_ring(F)
-  R = coefficient_ring(Rx)
-  ctx = MPolyBuildCtx(Rx)
-  
-  # shortcut in order not to allocate the dictionary
-  if isone(length(s))
-    (i, e, c) = first(s)
-    push_term!(ctx, R(c), e)
-    return FreeModElem(sparse_row(Rx, [(i, finish(ctx))]), F)
-  end
-
-  cache = IdDict{Int, typeof(ctx)}()
-  for (i, e, c) in s
-    ctx = get!(cache, i) do
-      MPolyBuildCtx(Rx)
-    end
-    push_term!(ctx, R(c), e)
-  end
-  return FreeModElem(sparse_row(Rx, [(i, finish(ctx)) for (i, ctx) in cache]), F)
+  row = _build_sparse_row(Rx, s)
+  return FreeModElem(row, F)
 end
 
 function (F::FreeMod{<:MPolyQuoRingElem})(s::Singular.svector)
+  is_zero(s) && return zero(F) # `map_entries` throws for empty vectors
   Qx = base_ring(F)::MPolyQuoRing
   Rx = base_ring(Qx)::MPolyRing
+  row = _build_sparse_row(Rx, s)
+  return FreeModElem(map_entries(Qx, row), F)
+end
+
+function _build_sparse_row(Rx::Ring, s::Singular.svector)
   R = coefficient_ring(Rx)
   ctx = MPolyBuildCtx(Rx)
 
-  # shortcut in order not to allocate the dictionary
+  # shortcuts in order not to allocate the dictionary
+  is_zero(length(s)) && return sparse_row(Rx)
   if isone(length(s))
     (i, e, c) = first(s)
     push_term!(ctx, R(c), e)
-    return FreeModElem(sparse_row(Qx, [(i, Qx(finish(ctx)))]), F)
+    return sparse_row(Rx, [(i, finish(ctx))])
   end
 
   cache = IdDict{Int, typeof(ctx)}()
```

Comment:
```suggestion
  cache = Dict{Int, typeof(ctx)}()
```
A `Dict` is faster if the keys are are of type `Int`.

Reply:
Oh! I didn't know. Thanks.